File: solution/22/solution.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        if n==0:return result
        generate('(', 1, 0, n, result)
        return result

    def generate(item, left_num, right_num, n, result):
        if left_num == right_num == n:
            result.append(item)
            return
        if left_num > n or right_num > left_num:
            return
        left_num += 1
        generate(item+'(', left_num, right_num, n, result)
        left_num -= 1
        right_num += 1
        generate(item+')', left_num, right_num, n, result)

File: solution/22/solution.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        if(n == 0)return result;
        string item = "(";
        int left_num = 1;
        int right_num = 0;
        generate(result, "(", n, left_num, right_num);
        return result;
    }

    void generate(vector<string>& result, string item,int n, int left_num, int right_num){
        if(left_num == n&&right_num == n){
            result.push_back(item);
            return;
        }
        if(left_num>n||right_num>left_num)return;
        left_num ++;
        generate(result, item+'(', n, left_num, right_num);
        left_num --;
        right_num ++;
        generate(result, item+')', n, left_num, right_num);
    }
};